sql/spanner/inspect.go

```diff
@@ -0,0 +1,738 @@
+
+	switch col.typ {
+	case TypeInt64:
+		typ = &schema.IntegerType{T: col.typ}
+	case TypeBool:
+		typ = &schema.BoolType{T: col.typ}
+	case TypeTimestamp:
+		typ = &schema.TimeType{T: col.typ}
+	case TypeDate:
+		typ = &schema.TimeType{T: col.typ}
+	case TypeJSON:
+		typ = &schema.JSONType{T: col.typ}
+	case TypeNumeric:
+		typ = &schema.DecimalType{T: col.typ}
+	default:
+		if strings.HasPrefix(col.typ, TypeString) {
+			typ = &schema.StringType{
+				T:    col.typ,
+				Size: col.size,
+				Attrs: []schema.Attr{
+					&SizeIsMax{Bool: col.sizeIsMax},
+				},
+			}
+		} else if strings.HasPrefix(col.typ, TypeBytes) {
+			typ = &schema.BinaryType{
+				T:    col.typ,
+				Size: &col.size,
+				Attrs: []schema.Attr{
+					&SizeIsMax{Bool: col.sizeIsMax},
+				},
+			}
+		} else {
+			typ = &schema.UnsupportedType{T: col.typ}
+		}
+	}
+	return typ, nil
+}
+
+// indexes queries and appends the indexes of the given table.
+func (i *inspect) indexes(ctx context.Context, s *schema.Schema) error {
+	rows, err := i.querySchema(ctx, indexesQuery, s)
+	if err != nil {
+		return fmt.Errorf("spanner: querying schema %q indexes: %w", s.Name, err)
+	}
+	defer rows.Close()
+	if err := i.addIndexes(s, rows); err != nil {
+		return err
+	}
+	return rows.Err()
+}
+
+// addIndexes scans the rows and adds the indexes to the table.
+func (i *inspect) addIndexes(s *schema.Schema, rows *sql.Rows) error {
+	names := make(map[string]*schema.Index)
+	for rows.Next() {
+		var (
+			tableSchema                     sql.NullString
+			tableName, indexName, indexType string
+			parentTableName                 sql.NullString
+			isUnique, isNullFiltered        sql.NullBool
+			columnName                      sql.NullString
+			ordinalPosition                 sql.NullInt64
+			columnOrdering                  sql.NullString
+			isNullable                      sql.NullBool
+		)
+		if err := rows.Scan(
+			&tableSchema, &tableName, &indexName,
+			&indexType, &parentTableName, &isUnique,
+			&isNullFiltered, &columnName, &ordinalPosition,
+			&columnOrdering, &isNullable,
+		); err != nil {
+			return fmt.Errorf("spanner: scanning indexes for schema %q: %w", s.Name, err)
+		}
+		t, ok := s.Table(tableName)
+		if !ok {
+			return fmt.Errorf("table %q was not found in schema", tableName)
+		}
+		name := tableName + "_" + indexName
+		// Add Index if it doesn't exist.
+		idx, ok := names[name]
+		if !ok {
+			idx = &schema.Index{
+				Name:   indexName,
+				Unique: isUnique.Bool,
+				Table:  t,
+				Attrs: []schema.Attr{
+					&IndexType{T: indexType},
+				},
+			}
+			idx.AddAttrs(&IsNullFiltered{Bool: isNullFiltered.Bool})
+			names[name] = idx
+			if indexType == "PRIMARY_KEY" {
+				if t.PrimaryKey == nil {
+					t.PrimaryKey = idx
+				}
+			} else {
+				t.AddIndexes(idx)
+			}
+		}
+		// Add IndexColumnPart if it doesn't exist.
+		part := &schema.IndexPart{
+			Desc: columnOrdering.String == "DESC",
+			Attrs: []schema.Attr{
+				&PartOrdinal{Int64: ordinalPosition.Int64},
```

Comment:
`SeqNo` represents the position of the column in the index. It can be used here instead of the attribute.